pkgs/top-level/build-wrappers.nix

```diff
@@ -0,0 +1,113 @@
+# dependencies to inject. Care must be taken to avoid cycles--i.e. use unwrapped
+# or less-wrapped version of the tools when building the packages that are then
+# injected here.
+#
+# As an added bonus, since these wrappers are only available as part of
+# `buidlPackages`, it will be slightly harder to accidently depend on compilers
+# at runtime, which is usually a mistake.
+runPackages: self: super:
+
+with self;
+
+{
+  wrapCCWith = ccWrapper: libc: extraBuildCommands: baseCC: ccWrapper {
+    nativeTools = stdenv.cc.nativeTools or false;
+    nativeLibc = stdenv.cc.nativeLibc or false;
+    nativePrefix = stdenv.cc.nativePrefix or "";
+    cc = baseCC;
+    dyld = if stdenv.isDarwin then darwin.dyld else null;
+    isGNU = baseCC.isGNU or false;
+    isClang = baseCC.isClang or false;
+    inherit libc extraBuildCommands;
+  };
+
+  ccWrapperFun = callPackage ../build-support/cc-wrapper;
+
+  wrapCC = wrapCCWith ccWrapperFun runPackages.libc "";
+  # legacy version, used for gnat bootstrapping
+  wrapGCC-old = baseGCC: callPackage ../build-support/gcc-wrapper-old {
+    nativeTools = stdenv.cc.nativeTools or false;
+    nativeLibc = stdenv.cc.nativeLibc or false;
+    nativePrefix = stdenv.cc.nativePrefix or "";
+    gcc = baseGCC;
+    libc = glibc;
+  };
+
+  wrapGCCCross =
+    {gcc, libc, binutils, cross, shell ? "", name ? "gcc-cross-wrapper"}:
+
+    callPackage ../build-support/gcc-cross-wrapper {
+      nativeTools = false;
+      nativeLibc = false;
+      noLibc = (libc == null);
+      inherit gcc binutils libc shell name cross;
+    };
+
+  gccCrossStageStatic = assert stdenv ? cross; wrapGCCCross {
+    gcc = gccCrossStageStatic-unwrapped;
+    libc = libcTarget;
+    inherit binutils;
+    inherit (stdenv) cross;
+  };
+
+  # Only needed for mingw bootstrapping
+  gccCrossMingw2 = assert stdenv ? cross; wrapGCCCross {
+    gcc = gccCrossStageStatic-unwrapped;
+    libc = windows.mingw_headers2;
+    inherit binutils;
+    inherit (stdenv) cross;
+  };
+
+  gccCrossStageFinal = assert stdenv ? cross; wrapGCCCross {
+    gcc = gccCrossStageFinal-unwrapped;
+    inherit (runPackages) libc;
+    inherit binutils;
+    inherit (stdenv) cross;
+  };
+
+  gcc45 = wrapCC gcc45-unwrapped;
+  gcc48 = wrapCC gcc48-unwrapped;
+  gcc49 = wrapCC gcc49-unwrapped;
+  gcc5 = wrapCC gcc5-unwrapped;
+  gcc6 = wrapCC gcc6-unwrapped;
+  gcc = wrapCC gcc-unwrapped;
+
+  gcc_debug = wrapCC gcc_debug-unwrapped;
+
+  gfortran48 = wrapCC gfortran48-unwrapped;
+  gfortran49 = wrapCC gfortran49-unwrapped;
+  gfortran5 = wrapCC gfortran5-unwrapped;
+  gfortran6 = wrapCC gfortran6-unwrapped;
+
+  gcj49 = wrapCC gcj49-unwrapped;
+  gcj = wrapCC gcj-unwrapped;
+
+  gnat45 = wrapCC gnat45-unwrapped;
+  gnat = wrapCC gnat-unwrapped;
+
+  gnatboot = wrapGCC-old  gnatboot-unwrapped;
+
+  gccgo49 = wrapCC gccgo49-unwrapped;
+  gccgo = wrapCC gccgo-unwrapped;
+
+  clang_35 = wrapCC llvmPackages_35.clang;
+  clang_34 = wrapCC llvmPackages_34.clang;
```

Comment:
Are these linking against the headers provided with clang, or are these linking against the headers provided by gcc?
Note that gcc headers are not fully compatible with clang, because gcc supports more atomic builtins than clang, which cause compilation failures of Firefox.

https://github.com/mozilla/nixpkgs-mozilla/blob/master/release.nix#L51-L76